# Entanglement Teleportation and Quantum Network

We explore a protocol known as "entanglement teleportation", where two parties, Alice and Bob, each holding a $|\Phi^+\rangle$ Bell state, can entangle their qubits by sending 1 qubit from Alice to Bob and performing local operations. This can be very useful when entangled pairs need to be shared over very long distances and there is a limit on how far a qubit can be transmitted before it decoheres and its useful quantum information.

The basic setup is as follows: Alice and Bob want to share an entangled Bell pair so that they can perform other quantum protocols. Normally, Alice could just prepare a Bell pair $|\Psi^+\rangle$ and send one of the qubits over to Bob. Unfortunately, in this case, Alice and Bob are separated by a great distance. If Alice tried to send one of the qubits to Bob, it would decohere before it reaches Bob and the resulting pair would not be entangled. 

Thankfully, their mutual friend Charlie is located right between Alice and Bob, and Alice and Charlie are able to exchange entangled pairs without encountering decoherence (similarly, Charlie and Bob can exchange entangled pairs freely). Is there a way to "boost" the entangled state through Charlie?

## Entanglement Teleportation Protocol

We use the following notation to define the Bell states:
$$\begin{align*} \ket{\Phi^+} = \frac{1}{\sqrt{2}}(\ket{00}+\ket{11}), &\qquad \ket{\Phi^-} = \frac{1}{\sqrt{2}}(\ket{00}-\ket{11}) \\
 \ket{\Psi^+} = \frac{1}{\sqrt{2}}(\ket{01}+\ket{10}), &\qquad \ket{\Psi^-} = \frac{1}{\sqrt{2}}(\ket{01}-\ket{10}) \end{align*}$$
 
Alice and Charlie each start with two qubits: Alice with qubits $q_0q_1$ and Charlie with qubits $q_2q_3$. They prepare their qubits in a Bell state, and so Alice's initial state is $|\Phi^+\rangle_A$ and Charlie's is $|\Phi^+\rangle_C$. Now, the steps of the protocol are as follows:

1. Alice sends $q_1$ to Charlie
2. Charlie makes a measurement of $q_1q_2$ in the Bell basis and records the measurement outcomes $M\in\{M_{\Phi^+},M_{\Phi^-} M_{\Psi^+},M_{\Psi^-}\}$.
3. Based on the observed measurement outcome $M$, Charlie performs single-qubit operations on $q_3$.
The end result is that there is a Bell pair $|\Phi^+\rangle$ over $q_1q_2$, and another over $q_0q_3$: Alice and Charlie now share (two) entangled pairs of qubits!


This is pretty amazing, so let's take a closer look to see how it works.

### Step 1+2
Step 1 does nothing interesting to the quantum state (in fact we want it that way!) Let's look at what happens when Charlie makes a Bell-basis measurement. The state at the end of step 1 is as follows:
$$|\psi\rangle = \ket{\Phi^+}_A\ket{\Phi^+}_C = \left(\frac{1}{\sqrt{2}}(\ket{00}+\ket{11})_{01}\right)\left(\frac{1}{\sqrt{2}}(\ket{00}+\ket{11})_{23}\right) = \frac{1}{2}(\ket{0000}+\ket{0011}+\ket{1100}+\ket{1111})_{0123}$$

Where we use subscripts to indicate which qubit the ket represents (e.g.. $\ket{0000}_{0123}$ means the 0th through 3rd qubits are all 0). Let's rewrite the index 1 and 2 qubits in the Bell basis. Using identities such as $\ket{00} = \frac{1}{\sqrt{2}}(\ket{\Phi^+} + \ket{\Phi^-]})$, we have

$$\begin{align*}
        &\frac{1}{2}(\ket{0000} + \ket{0011} + \ket{1100} + \ket{1111})_{0123}\\
        &= \frac{1}{2\sqrt{2}}(\ket{0}_0(\ket{\Phi^+} +\ket{\Phi^-})_{12}\ket{0}_3 + \ket{0}_0(\ket{\Psi^+} + \ket{\Psi^-})_{12}\ket{1}_3 + \ket{1}_0(\ket{\Psi^+} -\ket{\Psi^-})_{12}\ket{0}_3 +\ket{1}_0(\ket{\Phi^+}-\ket{\Phi^-})_{12}\ket{1}_3)
    \end{align*}$$

This allows us to see what happens when Charlie performs the Bell-basis measurement on $q_1q_2$. When a particular outcome is measured, the superposition projects onto that outcome and is renormalized. Suppose Charlie measures $M_{\Phi^+}$ (ie $q_1q_2$ to be in the state $\ket{\Phi^+}$). To compute the resultant state, we apply the projector $\ket{\Phi^+}_{12}\bra{\Phi^+}$ to the initial state. Since the distinct Bell states are orthogonal to each other (check this!), we only keep terms that have $\ket{\Phi^+}$ in them, and we get:
$$\begin{align*} \ket{\psi^{(\Phi^+)}}&:=\ket{\Phi^+}_{12}\bra{\Phi^+}\frac{1}{2\sqrt{2}}(\ket{0}_0(\ket{\Phi^+} +\ket{\Phi^-})_{12}\ket{0}_3 + \ket{0}_0(\ket{\Psi^+} + \ket{\Psi^-})_{12}\ket{1}_3 + \ket{1}_0(\ket{\Psi^+} -\ket{\Psi^-})_{12}\ket{0}_3 +\ket{1}_0(\ket{\Phi^+}-\ket{\Phi^-})_{12}\ket{1}_3) \\
&= \frac{1}{\sqrt{2}}(\ket{0}_0\ket{\Phi^+}_{12}\ket{0}_3 + \ket{1}_0\ket{\Phi^+}_{12}\ket{1}_3)\\
&= \frac{1}{\sqrt{2}}(\ket{00}_{03}\ket{\Phi^+}_{12} + \ket{11}_{03}\ket{\Phi^+}_{12})\\
&= \ket{\Phi^+}_{03}\ket{\Psi^+}_{12}
\end{align*}
$$

The calculation shows that we still have two pair of entangled Bell states, but the entanglement is over a different pair of qubits! Notice how Alice's $q_0$ is now entangled with Charlie's $q_3$, while originally it was entangled with $q_1$.

Things are slightly different if Charlie measurement yields a different result. Without going into the details of the derivation, the resultant state after Charlie measures the other possible Bell states are as follow:
$$
\begin{align*}
\ket{\psi^{(\Phi^+)}}
        &= \ket{\Phi^-}_{03}\ket{\Phi^-}_{12} \\
        \ket{\psi^{(\Psi^+)}}
        &= \ket{\Psi^+}_{03}\ket{\Psi^+}_{12}\\
        \ket{\psi^{(\Psi^-)}}
        &= \ket{\Psi^-}_{03}\ket{\Psi^-}_{12}
\end{align*}
$$
The entanglement is still transferred to the pairs $q_0q_3$ and $q_1q_2$, but the Bell states are different! While this is technically fine for a 1-stage protocol, we need something better if we want to do muliple rounds of teleportation. Therefore we need the 3rd step.

### Step 3

In order for the protocol to be extensible, we want Charlie to be able to manipulate the qubits he has so that the entangled pair $q_0q_3$ is equal to $\ket{\Phi^+}$. But the Bell states have the property that each can be transformed into another with just a single-qubit gate. Specifically, we have
$$
\begin{align*}
Z_3\ket{\Phi^-}_{03} &= Z_3\frac{1}{\sqrt{2}}(\ket{00}-\ket{11})_{03} = \frac{1}{\sqrt{2}}(\ket{00}+\ket{11})_{03} = \ket{\Phi^+}_{03} \\
X_3\ket{\Psi^+}_{03} &= X_3\frac{1}{\sqrt{2}}(\ket{01}+\ket{10})_{03} = \frac{1}{\sqrt{2}}(\ket{00}+\ket{11})_{03} = \ket{\Phi^+}_{03} \\
Z_3X_3\ket{\Psi^-}_{03} &= Z_3X_3\frac{1}{\sqrt{2}}(\ket{01}-\ket{10})_{03} = Z_3\frac{1}{\sqrt{2}}(\ket{00}-\ket{11})_{03} = \ket{\Phi^+}_{03}
\end{align*}
$$
Therefore based on Charlie's measurements, he can apply $Z_3$ and/or $X_3$ gates to his qubit $q_3$ and transform the entangled pair between Alice and Charlie to become $\ket{\Phi^+}$.

## Implementation : Teleporting entanglement between Alice and Charlie

Let's start defining the relevant operators to implement this protocol. Firstly, we need a function that prepares the $\ket{\Phi^+}$ state from the initial all-zeros state:

In [1]:
operation PreparePhiPlus (x0 : Qubit, x1 : Qubit) : Unit is Adj + Ctl {
    H(x0);
    CX(x0, x1);
}

PreparePhiPlus

We also need to implement the Bell measurement as well as the subsequent single-qubit operations by Charlie. We will do both in the same function.
1. For the Bell basis measurement, we call the Adjoint of PreparePhiPlus to bring the Bell basis back into the standard basis. We can check how the states map:
$$
\begin{align*} 
H_1CX_{12}\ket{\Phi^+} &= \frac{1}{\sqrt{2}}H_1CX_{12}(\ket{00}+\ket{11})_{12} = \frac{1}{\sqrt{2}}H_1(\ket{00}+\ket{10})_{12} = \ket{00}_{12} \\
H_1CX_{12}\ket{\Phi^-} &= \frac{1}{\sqrt{2}}H_1CX_{12}(\ket{00}-\ket{11})_{12} = \frac{1}{\sqrt{2}}H_1(\ket{00}-\ket{10})_{12} = \ket{10}_{12} \\
H_1CX_{12}\ket{\Psi^+} &= \frac{1}{\sqrt{2}}H_1CX_{12}(\ket{01}+\ket{10})_{12} = \frac{1}{\sqrt{2}}H_1(\ket{01}+\ket{11})_{12} = \ket{01}_{12} \\
H_1CX_{12}\ket{\Psi^-} &= \frac{1}{\sqrt{2}}H_1CX_{12}(\ket{01}-\ket{10})_{12} = \frac{1}{\sqrt{2}}H_1(\ket{01}-\ket{11})_{12} = \ket{11}_{12}
\end{align*}
$$
Then using the measurement in the standard basis, we can figure out which Bell state the measurement corresponds to and apply the appropriate single-qubit gates to $q_3$ according to Step 3:

In [2]:
operation TeleportOnce (x0 : Qubit, x1: Qubit, x2: Qubit, x3: Qubit) : Unit {
    Adjoint PreparePhiPlus(x1, x2);
    if (M(x1) == One) {
        Z(x3);
    }
    if (M(x2) == One) {
        X(x3);
    }
}

TeleportOnce

Finally, we can write a simple function that implements the protocol for Alice and Charlie. We also write a function that repeats the protocol $n$ times and then collates the measurement results.

In [15]:
operation EntanglementSwapDemo() : Result[] {
    // Step 0 : Bell state initialization
    use (Alice, Charlie) = (Qubit[2], Qubit[2]);
    PreparePhiPlus(Alice[0], Alice[1]);
    PreparePhiPlus(Charlie[0], Charlie[1]);
    // Step 1+2+3 : Alice sends her second qubit to Charlie, who then does the Bell-measurement
    // and then appropriate single-qubit gates on his second qubit
    TeleportOnce(Alice[0], Alice[1], Charlie[0], Charlie[1]);
    // At this point we expect A0 C1 to be in the Bell state, so we convert it back to the standard basis and measure
    Adjoint PreparePhiPlus(Alice[0], Charlie[1]);
    Ignore(M(Alice[1]));
    Ignore(M(Charlie[0]));
    return [M(Alice[0]), M(Charlie[1])];
}   

operation RepeatDemo(n : Int) : Int[] {
    mutable counts = new Int[4];
    for i in 0..n-1 {
        let result = EntanglementSwapDemo();
        let value = (result[0] == One ? 1 | 0) + (result[1] == One ? 2 | 0);
        set counts w/= value <- counts[value]+1;
    }
    return counts;
}

c:\snippet_.qs(17,22): warning QS3308: Deprecated syntax. Use [] to construct an empty array, or [x, size = n] to construct an array of x repeated n times.


EntanglementSwapDemo, RepeatDemo

We should expect to see that the protocol successfully creates the Bell state $q_0q_3 = \ket{\Phi^+}$ 100% of the time - and it does!

In [14]:
%simulate RepeatDemo n=1000

1000, 0, 0, 0

## Finishing the protocol

That was very exciting, but we must not forget the original purpose of the protocol! We want to be able to form an entangled Bell pair between Alice and Bob. Before we continue, let's take a moment to recall the current state.

- Alice has qubit $q_0$ - which is currently entangled with Charlie's $q_3$ in the $\ket{\Phi^+}$ state.
- Charlie has qubits $q_1,q_2$ and $q_3$. $q_1q_2$ are no longer relevant to the protocol, and $q_3$ is entangled with Alice's $q_0$ as stated
- Bob has qubits $q_4q_5$ in the $\ket{\Phi^+}$ state.

Now it is Charlie's turn to play the role of Alice in this second half of the protocol. Everything is very similar to the first half:
1. Charlie sends $q_3$ to Bob
2. Bob makes a measurement of $q_3q_4$ in the Bell basis and records the measurement outcomes $M\in\{M_{\Phi^+},M_{\Phi^-} M_{\Psi^+},M_{\Psi^-}\}$.
3. Based on the observed measurement outcome $M$, Charlie performs single-qubit operations on $q_5$.
The end result is that $q_0q_5$ is now a Bell pair $\ket{\Phi^+}$ : Alice and Bob now share an entangled pair of qubits!

We will not do the full derivation here, but you should convince yourself that it works given how we analyzed Alice's and Charlie's part of the protocol.

## Quantum Repeater

In the above scenario, Charlie acted as a "quantum repeater" by helping Alice become entangled with Bob, something that would not have been possible if Alice tried to send her qubit $q_1$ to Bob directly (due to decoherence over long distances). This idea can be extended to improve the physical range of entanglement even further! If we had a long line of "Charlie"s between Alice and Bob, the entanglement teleportation protocol can be performed between Alice and Charlie #1, Charlie #1 and Charlie #2, and so on... all the way to Charlie #N and Bob, to establish an entangled pair between Alice and Bob.

# Testing on Quantum Hardware



With access to Microsoft Quantum Azure, we can implement the Entanglement Teleportation protocol and see how it performs on noisy devices! Firstly, let's write a function that will implement the protocol with $n$ people:

In [29]:
operation EntanglementSwap(n : Int) : Result[] {
    use q = Qubit[2*n];
    for i in 0..n-1 {
        PreparePhiPlus(q[2*i], q[2*i+1]);
    }
    for i in 0..n-2 {
        TeleportOnce(q[0], q[1+2*i], q[2+2*i], q[3+2*i]);
        Ignore(M(q[1+2*i]));
        Ignore(M(q[2+2*i]));
    }
    Adjoint PreparePhiPlus(q[0], q[2*n-1]);

    return [M(q[0]), M(q[2*n-1])];
} 

EntanglementSwap

We can see that setting $n=3$ gives us the entanglement teleportation protocol with Alice, Bob and Charlie!

Now we connect the the Azure workspace and check what devices we have access to:

In [18]:
%azure.connect "/subscriptions/b0c3bbe3-7123-4ba4-9aa2-57d9845f4a1d/resourceGroups/AzureQuantum/providers/Microsoft.Quantum/Workspaces/qchack-2022-b" location="East US"

Connected to Azure Quantum workspace qchack-2022-b in location eastus.


Target ID                   Current Availability Average Queue Time (Seconds)
--------------------------- ----------- ---
ionq.qpu                    Available   898
ionq.simulator              Available   2
quantinuum.hqs-lt-s1        Unavailable 0
quantinuum.hqs-lt-s1-apival Available   0
quantinuum.hqs-lt-s2        Unavailable 0
quantinuum.hqs-lt-s2-apival Available   0
quantinuum.hqs-lt-s1-sim    Available   0
quantinuum.hqs-lt-s2-sim    Available   7
quantinuum.hqs-lt           Unavailable 0

Because the protocol requires *classically*-controlled gates, we have to choose a backend that support those operations. Currently, the Ionq simulator and QPU do not support this type of algorithm. Therefore we will use the Quantinuum simulators and hardware. The Quantinuum simulators also simulate the noise that the QPU's gate operations induce, and so we will be able to learn how the entanglement teleportation algorithm performs on imperfect computers!

In [19]:
%azure.target quantinuum.hqs-lt-s2-sim

Loading package Microsoft.Quantum.Providers.Honeywell and dependencies...
Active target is now quantinuum.hqs-lt-s2-sim


Target ID                Current Availability Average Queue Time (Seconds)
------------------------ --------- -
quantinuum.hqs-lt-s2-sim Available 7

Before we run it on the backend, let's check once more that our protocol works by simulating it in-silico:

In [30]:
operation RepeatEntanglementSwap(n : Int, rep : Int) : Int[] {
    mutable counts = new Int[4];
    for i in 0..rep-1 {
        let result = EntanglementSwap(n);
        let value = (result[0] == One ? 1 | 0) + (result[1] == One ? 2 | 0);
        set counts w/= value <- counts[value]+1;
    }
    return counts;
}

c:\snippet_.qs(2,22): warning QS3308: Deprecated syntax. Use [] to construct an empty array, or [x, size = n] to construct an array of x repeated n times.


RepeatEntanglementSwap

In [34]:
%simulate RepeatEntanglementSwap n=3 rep = 1000

1000, 0, 0, 0

Great! Let's submit it to the simulator and wait for the results

In [35]:
%azure.submit EntanglementSwap n=3

Submitting EntanglementSwap to target quantinuum.hqs-lt-s2-sim...
Job successfully submitted for 500 shots.
   Job name: EntanglementSwap
   Job ID: 9d1505b7-9548-4ce8-84ff-9b55e63e9330


Job Name         Job ID                                                                                                                                                                                                                                                                                                                                                 Job Status Target                   Creation Time               Begin Execution Time End Execution Time Cost Estimate
---------------- ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ ------- ------------------------ ---------------------------
EntanglementSwap <a href="https://portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/b0c3bbe3-7123-4ba4-9aa2-57d9845f4a1d/resourceGroups/AzureQuantum/providers/Microsoft.Quantum/Workspaces/qchack-2022-b/job_management?microsoft_azure_quantum_jobid=9d1505b7-9548-4ce8-84ff-9b55e63e9330" target="_blank">9d1505b7-9548-4ce8-84ff-9b55e63e9330</a> Waiting quantinuum.hqs-lt-s2-sim 10/4/2022 2:42:51 AM +00:00

In [36]:
%azure.status

Job Name         Job ID                                                                                                                                                                                                                                                                                                                                                 Job Status Target                   Creation Time               Begin Execution Time        End Execution Time          Cost Estimate
---------------- ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ --------- ------------------------ --------------------------- --------------------------- ---------------------------
EntanglementSwap <a href="https://portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/b0c3bbe3-7123-4ba4-9aa2-57d9845f4a1d/resourceGroups/AzureQuantum/providers/Microsoft.Quantum/Workspaces/qchack-2022-b/job_management?microsoft_azure_quantum_jobid=9d1505b7-9548-4ce8-84ff-9b55e63e9330" target="_blank">9d1505b7-9548-4ce8-84ff-9b55e63e9330</a> Succeeded quantinuum.hqs-lt-s2-sim 10/4/2022 2:42:51 AM +00:00 10/4/2022 2:44:02 AM +00:00 10/4/2022 2:44:02 AM +00:00

Once the above cell reports that the job has succeeded, we can print out the histogram of results below:

In [37]:
%azure.output

Result Frequency
----- --------------------
[0,0] 0.9400000000000007
[0,1] 0.02800000000000001
[1,1] 0.008
[1,0] 0.024000000000000007

Pretty good! Seems like the fidelity of the teleportation is around $93\sim95\%$

Now lets try a longer protocol, this time between 5 parties:

In [38]:
%azure.submit EntanglementSwap n=5

Submitting EntanglementSwap to target quantinuum.hqs-lt-s2-sim...
Job successfully submitted for 500 shots.
   Job name: EntanglementSwap
   Job ID: f44fe681-a577-47e2-9da6-837b95a8629d


Job Name         Job ID                                                                                                                                                                                                                                                                                                                                                 Job Status Target                   Creation Time               Begin Execution Time End Execution Time Cost Estimate
---------------- ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ ------- ------------------------ ---------------------------
EntanglementSwap <a href="https://portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/b0c3bbe3-7123-4ba4-9aa2-57d9845f4a1d/resourceGroups/AzureQuantum/providers/Microsoft.Quantum/Workspaces/qchack-2022-b/job_management?microsoft_azure_quantum_jobid=f44fe681-a577-47e2-9da6-837b95a8629d" target="_blank">f44fe681-a577-47e2-9da6-837b95a8629d</a> Waiting quantinuum.hqs-lt-s2-sim 10/4/2022 2:45:14 AM +00:00

In [40]:
%azure.status

Job Name         Job ID                                                                                                                                                                                                                                                                                                                                                 Job Status Target                   Creation Time               Begin Execution Time        End Execution Time          Cost Estimate
---------------- ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ --------- ------------------------ --------------------------- --------------------------- ---------------------------
EntanglementSwap <a href="https://portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/b0c3bbe3-7123-4ba4-9aa2-57d9845f4a1d/resourceGroups/AzureQuantum/providers/Microsoft.Quantum/Workspaces/qchack-2022-b/job_management?microsoft_azure_quantum_jobid=f44fe681-a577-47e2-9da6-837b95a8629d" target="_blank">f44fe681-a577-47e2-9da6-837b95a8629d</a> Succeeded quantinuum.hqs-lt-s2-sim 10/4/2022 2:45:14 AM +00:00 10/4/2022 2:46:53 AM +00:00 10/4/2022 2:46:57 AM +00:00

In [41]:
%azure.output

Result Frequency
----- --------------------
[0,0] 0.9380000000000007
[1,1] 0.006
[1,0] 0.02600000000000001
[0,1] 0.030000000000000013

We see that the fidelity of the desired state decreased sligtly, but is still pretty high! 

Let's increase the number of parties by 1 again and see how well the protocol performs this time:

In [45]:
%azure.submit EntanglementSwap n=6

Submitting EntanglementSwap to target quantinuum.hqs-lt-s2-sim...
Job successfully submitted for 500 shots.
   Job name: EntanglementSwap
   Job ID: fb8e1d58-5225-4d79-bcd0-ce2baaaada9c


Job Name         Job ID                                                                                                                                                                                                                                                                                                                                                 Job Status Target                   Creation Time               Begin Execution Time End Execution Time Cost Estimate
---------------- ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ ------- ------------------------ ---------------------------
EntanglementSwap <a href="https://portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/b0c3bbe3-7123-4ba4-9aa2-57d9845f4a1d/resourceGroups/AzureQuantum/providers/Microsoft.Quantum/Workspaces/qchack-2022-b/job_management?microsoft_azure_quantum_jobid=fb8e1d58-5225-4d79-bcd0-ce2baaaada9c" target="_blank">fb8e1d58-5225-4d79-bcd0-ce2baaaada9c</a> Waiting quantinuum.hqs-lt-s2-sim 10/4/2022 2:56:21 AM +00:00

In [46]:
%azure.output

Result Frequency
----- --------------------
[0,0] 0.8720000000000007
[1,1] 0.018000000000000002
[1,0] 0.07000000000000005
[0,1] 0.04000000000000002

Wow, the circuit depth and resultant noise has taken its toll on the fidelity of the result! With around an $87%$ fidelity, this means that there is a 1 in 8 chance that the Alice and Bob do not share an entangled state at the end of the teleportation protocol.

Unfortunately the Quantinuum backend cannot support $n>6$, but we can already begin to see that gate noise means that there is an upper limit to how extensible the entanglement protocol is, even though in theory it works 100% of the time. In the future, hopefully the gate fidelity improves to the point where we can build such quantum networks across dozens of parties while maintaining extremely high fidelity!